iotile_analytics_interactive/iotile_analytics/interactive/reports/stream_report.py

```diff
@@ -0,0 +1,113 @@
+
+    This report generates two plots.  The first is a bar graph that shows how many
+    data points were received in this stream every X hours/weeks/days etc.  This is
+    particularly useful for diagnosing issue where data should be received regularly but
+    is not.  It can also help you orient yourself when looking through a stream with
+    data collected over widely separated time intervals like data for 2 months then a 3 month
+    gap then 1 month of data, etc.
+
+    The second plot is just a graph of the data in the stream over time.  You units of the
+    graph are whatever the internal units of the stream are unless you override them by
+    passing the name of your desired units as a parameter.  If you do choose to override
+    the units, you must specify a units name that has a known conversion from the internal
+    stream units.
+
+    Args:
+        stream (str): A unique identifier for the stream that we wish to download
+            and display.  This can be anything that could be passed to AnalysisGroup.fetch_stream
+            including a 4-character variable ID, a partial string that matches the label of the
+            stream, etc.
+        window (str): The aggregation period for showing when there is data in a stream.  You can
+            pass a value like 'days', 'hours', 'weeks', 'months'.  This will be given to TimeSelectViewer
+            so any valid value there is acceptable here.  Defaults to days.
+        units (str): The desired units in which you want to see the data.  This must be defined as an
+            available output unit on the stream.  If not passed, the default is to use whatever the
+            internal units are of the stream.  You can see what output unit options are by passing
+            a random string here and looking at the error message.
+        mdo (list(float)): An optional MDO to use to convert the data stored in this stream.  If passed
+            this must be 3 float values that define a linear transformation of the data stored in the stream.
+            The values are interpreted as m/d*value + o where the argument is specified as [m, d, o].
+    """
+
+    def __init__(self, group, stream, window='days', units=None, mdo=None):
+        super(StreamOverviewReport, self).__init__(self.UNHOSTED)
+
+        self.stream_data = group.fetch_stream(stream)
+        self.units = units
+        self.mdo = mdo
+        if units is not None:
+            if units not in self.stream_data.available_units:
+                raise UsageError("Unknown output units specified", known_units=self.stream_data.available_units)
+
+            self.stream_data = self.stream_data.convert(units)
+        if mdo is not None:
+            if len(mdo) != 3:
+                raise UsageError("When specifying a fixed MDO, you must specify 3 float values for the multiple, divide and offset", mdo=mdo)
+
+            self.stream_data = self.stream_data.apply_mdo(*mdo)
```

Comment:
Isn't the MDO going to be tricky for the user? As the `/df` already applies an MDO, so I am guessing this MDO here is after that MDO.

If we want to be able to overwrite it, we may want to add MDO as an option to the /df API so the server applies that MDO

Reply:
Just to clarify this, my tests show `/df/` returns the interval value of the stream without applying output unit MDO.  I hope this was intended because it is the behavior that `iotile_analytics` needs since it needs to be able to choose different output units so it prefers to get the internal data and apply the output transformation itself.

I don't envision users applying MDO transformations themselves since they should just use the `--units` argument to specify the units they want by name but it is important for system streams like `internal_temperature` and `battery_voltage` that are internally stored in non-user-friendly units and there is no associated stream object that defines nice unit names.  This was intended to just be an escape hatch to enable us to run reports on internal variables that doesn't have stream objects

Reply:
There definitely needs to be an option to get the internal value without any output units applied.  It's not about what units the user needs to see so much in a report as about being able to write numerical processing routines that know what their input units are and thus know what the algorithm is going to calculate.  It would be much harder if no part of `iotile-analytics` could ever reason about what units it was going to get from a stream because the user may have wanted to see a graph in different units someplace out.

So if the `/df/` behavior is going to change to include output units by default, I'll also need a flag to turn that off so I can update `iotile-analytics` at the same time because all of the reports and associated calculations assume their input data units are known and unchangeable.

Reply:
It could also be that every stream I've tried has had default output units applied that are equivalent to their internal units but if that's the case, then I'll really need an option so I can update `iotile-analytics` to force internal units mode.

Also, this API seems to be about ~5-10x faster than using `/data/?filter=<stream>` which is why I started moving to it.  This is especially important for streams that have large numbers of data points.